src/backend/cdb/cdbpartition.c

```diff
@@ -4124,24 +4122,14 @@ selectListPartition(PartitionNode *partnode, Datum *values, bool *isnull,
 						 */
 
 						/*
-						 * The tupdesc tuple descriptor matches the table
-						 * schema, so it has the rule type
+						 * Type of rule is in fact the type of const inside the
+						 * rule object.
+						 * Type of expr is defined by attribute in tuple
+						 * descriptor whose number corresponds to current
+						 * partitioning key in partitioned table
 						 */
-						Oid			rhstypid = tupdesc->attrs[attno - 1]->atttypid;
-
-						/*
-						 * exprTypeOid is passed to us from our caller which
-						 * evaluated the expression. In some cases (e.g legacy
-						 * optimizer doing explicit casting), we don't compute
-						 * specify exprTypeOid. Assume lhstypid = rhstypid in
-						 * those cases
-						 */
-						Oid			lhstypid = exprTypeOid;
-
-						if (!OidIsValid(lhstypid))
-						{
-							lhstypid = rhstypid;
-						}
+						Oid	rhstypid = c->consttype;
+						Oid	lhstypid = tupdesc->attrs[attno - 1]->atttypid;
```

Comment:
Why `tupdesc->attrs[attno - 1]->atttypid` is now `lhstypid` instead of `rhstypid`?